docs/VirtualMachine.md

```diff
@@ -0,0 +1,160 @@
+# Virtual Machine Design
+
+This document details the workings and design of the virtual machine that executes a Flux script.
+
+## Execution Stack
+
+A Flux script goes through several transformations from text to execution.
+
+1. Flux source code is created via any means.
+2. The parser produces an abstract syntax tree (AST) from the source code.
+3. The semantic analyzer produces a semantic graph from the AST.
+4. The compiler accepts the semantic graph and produces an intermediate representation (IR) to be executed.
+5. The flux virtual machine executes the IR.
+
+NOTE: Should we call it bytecode or IR?
+Should we call it a VM or an interpreter?
+
+## Execution model
+
+Flux has a specific execution model.
+The model depends on the definitions of various data types and operations on those types.
+The Flux spec defines some of those types.
+This document reviews those types as well as defines a few new types that are specific to the implementation.
+
+### Data types and operations
+
+#### Table
+
+A Flux table consists of rows and columns.
+Columns have a label and a type.
+The table has zero or more rows, where each row defines a value for each column or a null is used to indicate a missing value.
+
+A table additionaly has a group key, which is a list of columns and their values.
+All rows within a table have the same value for all columns that are part of the group key.
+
+#### Stream
+
+A stream is an unbounded set of tables.
+Each table must have a different group key.
```

Comment:
Is this strictly true? If the stream is unbounded, you could have the same table/schema/group key come up multiple times during different time frames. Like if you have group keys A & B and you have data that is streaming in from T 1-10, but you get in batches of 5. Then in the stream you'd have `A,T1-T5 : B,T1-T5 : A,T6-T10 : B,T6-T10`

Reply:
We do not support nor have we spec'ed out how unbounded streams will work. So according to the current Flux spec this is true. But as you say we will need to workout what this means for a streaming use case.

Reply:
conceptually, to get any meaningful output from a query on an unbounded stream there must be some checkpoint where in the context of the query checkpoint/trigger, all tables are finite and complete.  

then there are 2 definitions to think about  for an "unbounded"  stream.  
1.  unless the data itself defines it (e.g. all group keys are boolean)  you must assume that at any time a new table may appear.

2.  are aggregators, selectors, etc defined over the life of the query, since the last checkpoint, or something else completely?